fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java

```diff
@@ -355,7 +365,34 @@ public void afterSearch(FHIRPersistenceEvent event) throws FHIRPersistenceInterc
     }
 
     /**
-     * Enforce the authorizations granted by the end user in the form of scope strings
+     * Check whether the permissions required for the current interaction are granted by the approved scopes.
+     *
+     * @param resourceType the resource type used in the endpoint of the request (or "Resource" for whole-system interactions)
+     * @param requiredPermission the permission required for the requested interaction
+     * @param approvedScopes a list of SMART scopes associated with the request
+     * @throws FHIRPersistenceInterceptorException if the interaction is not permitted
+     */
+    private void checkScopes(String resourceType, Permission requiredPermission, List<Scope> approvedScopes) throws FHIRPersistenceInterceptorException {
+        for (Scope scope : approvedScopes) {
+            // "Resource" is used for "*" which applies to all resource types
+            if (scope.getResourceType() == ResourceType.Value.RESOURCE || scope.getResourceType().value().equals(resourceType)) {
+                if (hasPermission(scope.getPermission(), requiredPermission)) {
+                    return;
+                }
+            }
+        }
+
+        String msg = requiredPermission.value() + " permission for '" + resourceType +
+                "' is not granted by any of the provided scopes: " + approvedScopes;
+        if (log.isLoggable(Level.FINE)) {
+            log.fine(msg);
+        }
+        throw new FHIRPersistenceInterceptorException(msg)
```

Comment:
Can you use the same REQUEST_NOT_PERMITTED message that enforce() uses?

Reply:
I could have but I actually liked this one more because it gives the end user / app developer a better idea of why their request didn't go through (while still not leaking anything confidential.
I wouldn't mind making the more generic "Requested interaction is not permitted by any of the passed scopes." a little *less* generic, but for now I just focused on adding this new check.